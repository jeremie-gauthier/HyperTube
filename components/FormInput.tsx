import React from "react";
import classnames from "classnames";
import styles from "./FormInput.module.scss";

type FormInputProps = {
  value: string;
  name: string;
  placeholder: string;
} & Partial<DefaultProps>;

type DefaultProps = React.InputHTMLAttributes<HTMLInputElement> & {
  className: string;
  error: string;
};

export default function FormInput({
  className = "",
  error = undefined,
  ...rest
}: FormInputProps): JSX.Element {
  const [hasFocus, setHasFocus] = React.useState<boolean>(false);
  const hasValue = rest.value.length > 0;
  const showLabel = hasFocus || hasValue;

  const isFocused = classnames({
    [className]: true,
    "border-white": showLabel,
    "border-grey-light": !showLabel,
    [styles.inputError]: error,
  });

  return (
    <div className={styles.container}>
      {showLabel && <label htmlFor={rest.name}>{rest.placeholder}</label>}
      <input
        id={rest.name}
        {...rest}
        className={isFocused}
        placeholder={showLabel ? "" : rest.placeholder}
        onBlur={() => setHasFocus(false)}
        onFocus={() => setHasFocus(true)}
      />
      {error && <p>{error}</p>}
    </div>
  );
}
