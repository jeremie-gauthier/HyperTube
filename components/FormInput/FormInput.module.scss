.container {
  @apply relative;

  label {
    @apply absolute z-10 -top-2 left-2 px-1 text-xs rounded-b;

    // prettier-ignore
    background:
      linear-gradient(
        180deg,
        rgba(0, 0, 0, 0) 50%,
        rgba(65, 65, 65, 1) 50%
      );
  }

  input {
    @apply relative border rounded p-3 outline-none min-w-full;
  }

  .inputError {
    @apply border-b-2;

    border-bottom-color: rgb(229, 9, 20);
  }

  p {
    @apply text-red text-xs mt-1;
  }
}
