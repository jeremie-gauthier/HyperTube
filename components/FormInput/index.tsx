import React from "react";
import classnames from "classnames";
import { useTranslation } from "react-i18next";
import styles from "./FormInput.module.scss";

export type FormInputProps = {
  value: string;
  name: string;
  placeholder: string;
} & Partial<DefaultProps>;

type DefaultProps = React.InputHTMLAttributes<HTMLInputElement> & {
  error: string;
};

export default function FormInput({
  error = undefined,
  ...rest
}: FormInputProps) {
  const { hasFocus, handleFocus, handleBlur } = useFocus(
    rest.onFocus,
    rest.onBlur,
  );
  const hasValue = rest.value.length > 0;
  const showLabel = hasFocus || hasValue;
  const { t } = useTranslation();

  const isFocused = classnames({
    [rest.className ?? ""]: true,
    "border-white": hasFocus,
    "border-grey-light": !hasFocus,
    [styles.inputError]: error,
  });
  const labelStyle = classnames({
    "text-white": hasFocus,
    "text-grey-light": !hasFocus,
  });

  return (
    <div className={styles.container}>
      {showLabel && (
        <label htmlFor={rest.name} className={labelStyle}>
          {rest.placeholder}
        </label>
      )}
      <input
        id={rest.name}
        {...rest}
        className={isFocused}
        placeholder={showLabel ? "" : rest.placeholder}
        onBlur={handleBlur}
        onFocus={handleFocus}
        autoComplete="off"
      />
      {error && <p>{t(error)}</p>}
    </div>
  );
}

type FocusEvent =
  | ((evt: React.FocusEvent<HTMLInputElement>) => void)
  | undefined;

const useFocus = (onFocus: FocusEvent, onBlur: FocusEvent) => {
  const [hasFocus, setHasFocus] = React.useState(false);

  const handleFocus = React.useCallback(
    (evt) => {
      setHasFocus(true);
      if (onFocus) {
        onFocus(evt);
      }
    },
    [onFocus],
  );

  const handleBlur = React.useCallback(
    (evt) => {
      setHasFocus(false);
      if (onBlur) {
        onBlur(evt);
      }
    },
    [onBlur],
  );

  return { hasFocus, handleFocus, handleBlur };
};
