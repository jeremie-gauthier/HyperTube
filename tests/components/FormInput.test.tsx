import React from "react";
import FormInput from "@/components/FormInput";
import renderer from "react-test-renderer";
import { fireEvent, render } from "@testing-library/react";

test("<FormInput /> should render correctly without error", () => {
  const tree = renderer
    .create(<FormInput value="" name="test" placeholder="test" />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("<FormInput /> should render correctly with error", () => {
  const tree = renderer
    .create(
      <FormInput
        value=""
        name="test"
        placeholder="test"
        error="something wrong, I can feel it"
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("<FormInput /> should render correctly when focused", () => {
  const screen = render(
    <FormInput value="" name="test" placeholder="test" onChange={() => null} />,
    {},
  );
  const input = screen.getByPlaceholderText("test");
  fireEvent.focus(input);
  expect(screen.asFragment()).toMatchSnapshot();
  fireEvent.blur(input);
  expect(screen.asFragment()).toMatchSnapshot();
});
