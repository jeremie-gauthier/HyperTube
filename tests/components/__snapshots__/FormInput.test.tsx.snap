// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormInput /> should render correctly when focused 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <label
      class="text-white"
      for="test"
    >
      test
    </label>
    <input
      autocomplete="off"
      class=" border-white"
      id="test"
      name="test"
      placeholder=""
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`<FormInput /> should render correctly when focused 2`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <input
      autocomplete="off"
      class=" border-grey-light"
      id="test"
      name="test"
      placeholder="test"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`<FormInput /> should render correctly with error 1`] = `
<div
  className="container"
>
  <input
    autoComplete="off"
    className=" border-grey-light inputError"
    id="test"
    name="test"
    onBlur={[Function]}
    onFocus={[Function]}
    placeholder="test"
    value=""
  />
  <p>
    something wrong, I can feel it
  </p>
</div>
`;

exports[`<FormInput /> should render correctly without error 1`] = `
<div
  className="container"
>
  <input
    autoComplete="off"
    className=" border-grey-light"
    id="test"
    name="test"
    onBlur={[Function]}
    onFocus={[Function]}
    placeholder="test"
    value=""
  />
</div>
`;
